detect empty variable names

shaomeng opened the pull request:
This PR makes FlowRenderer detect user selections of variable names, and return an error code in the case of empty variable name. 
It would fix #1866 .

clyne commented on lib/render/FlowRenderer.cpp:
```diff
@@ -450,13 +454,26 @@ FlowRenderer::_updateFlowCacheAndStates( const FlowParams* params )
     // If names not the same, entire stream is out of date
     // Note: variable names are kept in VaporFields.
     std::vector<std::string> varnames = params->GetFieldVariableNames();
+    for( const auto& e : varnames )
```
I'm not sure that this code does what we want. It should not be an error for a user to select fewer than three variables. The missing variables should be treated as the "zero" variable: a variable that contains all zeros. A common use case is to do flow advection using only the horizontal components of velocity and ignore the vertical.

shaomeng replied:
Oh really? That's interesting, and I I'll change the behavior of it.

shaomeng commented:
@clyne , this PR is modified so that the behavior is now the same as what you described. It's time to review the changes.

clyne commented on include/vapor/ConstantGrid.h:
```diff
@@ -0,0 +1,85 @@
+{
+public:
+
+    // The constant value is specified via the constructor
```
The constant grid is also a dimensionless grid. I wonder if that is what we want, or if it would be better to specify dimensions, and coordinates? Maybe instead of being derived from Grid it should be derived from StructuredGrid (or RegularGrid). Something to consider for the future if we use ConstantGrid elsewhere.

shaomeng replied:
I guess we can have create derivatives of `ConstantGrid` that represent `ConstantStructuredGrid` and `ConstantRegularGrid`, and each one of them will have more specific information held by them?

clyne replied:
At some point we want to make ConstantGrid more like a real Grid class: have all of the public methods do reasonable things so that a ConstantGrid instance can easily be used anywhere we are using a Grid class. For now I'd say let's just move on since the current implementation is meeting our needs for flow visualization.

clyne commented on lib/flow/VaporField.cpp:
```diff
@@ -156,18 +174,28 @@ VaporField::GetVelocity( float time, const glm::vec3& pos, glm::vec3& velocity,
+            }
+            else
+            {
+                grid = _getAGrid( currentTS, varname );
```
How about having _getAGrid return a constant grid as needed and eliminating all of these conditionals?

shaomeng replied:
fixed.

clyne commented on lib/flow/VaporField.cpp:
```diff
@@ -124,12 +133,21 @@ VaporField::GetFirstStepVelocityIntersection( glm::vec3& minxyz, glm::vec3& maxx
+        }
+        else
+        {
+            grid = _getAGrid( 0, varname );
```
Here is an instance where if ConstantGrid were an actual grid with dimensions and coordinates this code wouldn't require special handling (conditionals)

shaomeng replied:
I changed the behavior of `GetUserExtents()` to return floating point mins and maxes. I guess if we decided to further implement things like `ConstantStructuredGrid`, then it would make sense for those derived classes to keep an actual user extent.

clyne commented on lib/vdc/ConstantGrid.cpp:
```diff
@@ -0,0 +1,104 @@
+    return 0;
+}
+    
+VAPoR::Grid::ConstCoordItr ConstantGrid::ConstCoordBegin() const
```
Probably should throw an assert here. What will happen if someone tries to use the returned iterator?

shaomeng replied:
Well, I moved these functions that are not supposed to be called by a user into a private block. In that way, it's guaranteed that nobody would call them.

clyne replied:
ok

clyne commented on lib/vdc/ConstantGrid.cpp:
```diff
@@ -0,0 +1,104 @@
+    
+float ConstantGrid::GetValueNearestNeighbor( const std::vector <double> &coords) const
+{
+    return 0.0f;
```
Should this return _value instead of 0.0?

shaomeng replied:
fixed.

clyne commented on lib/vdc/ConstantGrid.cpp:
```diff
@@ -0,0 +1,104 @@
+    
+float ConstantGrid::GetValueLinear( const std::vector <double> &coords) const 
+{
+    return 0.0f;
```
Return _value?

shaomeng replied:
fixed.

shaomeng commented:
Hey @clyne , I've updated the behavior of `ConstantGrid` to address some of your comments. You can see the updated header file to have an idea on what I've changed. 

If you also think we should go ahead and implement things like `ConstantRegularGrid`, then please let me know!
